### PR TITLE
Updates to developer guidelines

### LINKED ISSUE #371: Add 'how to write an application' to the development section of the documentation
This should include maybe an example skeleton application.

### PR DESCRIPTION
**Pull request by GernotMaier**
This PR includes updates to the *Guidelines for Developers*, see the (old page) https://gammasim.github.io/gammasim-tools/guidelines.html

I don't know if there is a possibility to see the changes in html.

Closes #371

### CODE REVIEW

File: docs/source/guidelines.rst [hunk 1]
```diff
@@ -1,48 +1,200 @@
 .. _Guidelines:
 
-Guidelines for Developers
-*************************
+Guidelines for Gammasim-tools Developers
```
Comment (orelgueta) [inline docs/source/guidelines.rst:3]: We never wrote `Gammasim-tools` with `Gamma` capitalised. Everywhere else it is `gammasim-tools`. If you don't mind, I would modify it here to be `gammasim-tools`.

File: docs/source/guidelines.rst [hunk 2]
```diff
@@ -1,48 +1,200 @@
...
+****************************************
 
-This section is meant for developers.
+This section provides help and guidelines for developers of Gammasim-tools.
+If you want to contribute to Gammasim-tools, please use one of the contact points listed at the entry page of this documentation.
+In general, please take note of the `ctapipe Development Guidelines <https://cta-observatory.github.io/ctapipe/development/index.html>`_.
```
Comment (orelgueta) [inline docs/source/guidelines.rst:8]: If you don't mind I will add explicitly a mention here that these guidelines are from ctapipe.

File: docs/source/guidelines.rst [hunk 3]
```diff
@@ -1,48 +1,200 @@
...
-Documentation
+Project setup
 =============
 
-Sphinx is used to create this documentation automatically with each merge into the master branch, see the `Github Action workflow CI-docs <https://github.com/gammasim/gammasim-tools/blob/master/.github/workflows/CI-docs.yml>`_.
+The main code repository for Gammasim-tools is on GitHub:
+
+`https://github.com/gammasim/gammasim-tools <https://github.com/gammasim/gammasim-tools>`_
+
+The main folders for developers are the `simtools <https://github.com/gammasim/gammasim-tools/tree/master/simtools>`_, `applications <https://github.com/gammasim/gammasim-tools/tree/master/applications>`_, `tests <https://github.com/gammasim/gammasim-tools/tree/master/tests>`_, and `docs <https://github.com/gammasim/gammasim-tools/tree/master/docs>`_ folders.
```
Comment (orelgueta) [inline docs/source/guidelines.rst:18]: I will also change `folders` to `directories` (`folder` is Windows speak, which I know you love...)

File: docs/source/guidelines.rst [hunk 4]
```diff
@@ -1,48 +1,200 @@
...
+==============
+
+Gammasim-tools is currently developed for Python 3.9.
+
+
+Code formatting
+===============
 
-For testing of Sphinx:
+Linting and code checks are done automatically using a CI workflow using ``pyflakes`` (see this `GitHub action workflow <https://github.com/gammasim/gammasim-tools/blob/master/.github/workflows/CI-unittests.yml>`_) and the pre-commit functionality.
```
Comment (orelgueta) [inline docs/source/guidelines.rst:30]: The CI unit tests don't really do code checks, right? I think I would mention just the pre-commit functionality (mentioning also black, sort, etc.) and then mention that as part of the CI also Codacy checks code formatting.

File: docs/source/guidelines.rst [hunk 5]
```diff
@@ -1,48 +1,200 @@
...
+
+    git commit --no-verify
+
+Logging
+=======
+
+Sufficient logging information should be provided to users and developers. As general guideline, the following logging levels should be used:
+
+- **INFO**: information useful for the general user about the progress, intermediate results, input or output.
+- **WARNING**: information for the general user or developer on something she should know but cannot change.
```
Comment (orelgueta) [inline docs/source/guidelines.rst:58]: Would change `she` to `they`

File: docs/source/guidelines.rst [hunk 6]
```diff
@@ -1,48 +1,200 @@
...
+
+For writing and testing documentation locally:
+
+.. code-block::
+
+    cd docs
+    make html
+
+This is especially recommended to identify warnings and errors by Sphinx (e.g., from badly formatted docstrings or RST files).
+The documentation can be viewed locally in a browser starting from the file ``./docs/build/html/index.html``.
```
Comment (orelgueta) [inline docs/source/guidelines.rst:128]: Technically you are already in the `docs` directory, no?

File: docs/source/guidelines.rst [hunk 7]
```diff
@@ -1,48 +1,200 @@
...
+
+
+Writing Applications
+====================
 
-For a reference of the numpydoc names, see https://numpydoc.readthedocs.io/en/latest/format.html
+Applications are command lines tools that should be build off of the simtools library.
```
Comment (orelgueta) [inline docs/source/guidelines.rst:134]: build --> built

File: docs/source/guidelines.rst [hunk 8]
```diff
@@ -1,48 +1,200 @@
...
+============
+
+Dependencies on python packages are listed in the `environment file <https://github.com/gammasim/gammasim-tools/blob/master/environment.yml>`_.
+Some of the packages installed are used for the development only and not needed for executing Gammasim-tools applications.
+
+
+Integration with CORSIKA and sim_telarray
+=========================================
+
+CORSIKA and sim_telarray are external tools to simtools.
```
Comment (orelgueta) [inline docs/source/guidelines.rst:176]: simtools --> gammasim-tools

File: docs/source/guidelines.rst [hunk 9]
```diff
@@ -58,19 +210,19 @@ For example:
...
 
 Validating names
-================
+----------------
 
-Any name that is recurrently used along the the package should be validated when given as input.
+Names that are recurrently used along the the package should be validated when given as input.
 Examples of names are: telescope, site, camera, model version. The functionalities to validate names
-are found in util.names. The function validateName receives the input string and a name dictionary,
+are found in  :ref:`util.names <utilnames>`. The function validateName receives the input string and a name dictionary,
```
Comment (orelgueta) [inline docs/source/guidelines.rst:225]: validateName --> validate_name

File: docs/source/guidelines.rst [hunk 10]
```diff
@@ -58,19 +210,19 @@ For example:
...
 that is usually called allSomethingNames. This dictionary contain the possible names (as keys) and lists
```
Comment (orelgueta) [inline docs/source/guidelines.rst:226]: allSomethingNames --> all_something_names

File: docs/source/guidelines.rst [hunk 11]
```diff
@@ -156,55 +296,31 @@ properties. See an example below:
 * names is a list of acceptable input names. The key in the returned dict will have the name given at the definition of the block (zenithAngle in this example)
```
Comment (orelgueta) [inline docs/source/guidelines.rst:296]: zenithAngle --> zenith_angle

File: docs/source/guidelines.rst [hunk 12]
```diff
@@ -156,55 +296,31 @@ properties. See an example below:
...
 
-One example of this approach is shower_simulator module, that is an interface module that
-connects to the tool specific module that is meant to manage shower simulations. In this case,
-this tool specific module is the corsika_runner, which can be replaced in the future.
+    # create a working directory
+    mkdir external && cd external
+    # clone gammasim-tools repository
+    git clone https://github.com/gammasim/gammasim-tools.git
+    # startup a container (download if is not available in your environment)
+    docker run --rm -it -v "$(pwd)/external:/workdir/external" ghcr.io/gammasim/containers/gammasim-tools-dev:v0.3.0-dev1 bash -c "$(cat ./entrypoint.sh) && bash"
```
Comment (orelgueta) [inline docs/source/guidelines.rst:323]: Is `entrypoint.sh` provided with gammasim-tools or should it be given here in the doc as well?

**Review by orelgueta — COMMENTED** [review]
A few minor comments. I can make all of the changes mentioned myself, but I just want your OK on the non-trivial ones.

File: docs/source/guidelines.rst (hunk 3 above)
Comment (GernotMaier) [reply docs/source/guidelines.rst:18]: I will get my Windows laptop today! Super excited!

**Comment by GernotMaier** [comment]
@orelgueta - please go ahead and add your changes.

**Comment by orelgueta** [comment]
OK, I will do it. What about the question regarding `entrypoint.sh`?

**Comment by GernotMaier** [comment]
I thought I’ve answered it - entrypoint.sh is part of the container image and called inside the container. So it is there.

> On 17. Nov 2022, at 09:52, orelgueta ***@***.***> wrote:
> 
> 
> OK, I will do it. What about the question regarding entrypoint.sh?
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were assigned.Message ID: ***@***.***>



---
Dr Gernot Maier
CTA
Deutsches Elektronen-Synchrotron DESY 
Ein Forschungszentrum der Helmholtz-Gemeinschaft
Platanenallee 6, D-15738 Zeuthen, Germany

Tel.: +49  33 7627 7598
Besucheradresse: 1X07

**Review by orelgueta — APPROVED** [review]
Fixed my comments (and a few other stuff) and approving. Merge if you agree with the changes.

**Comment by GernotMaier** [comment]
All good, thank you!

**Pull request merged**